server/src/main/java/org/elasticsearch/index/mapper/Mapper.java

```diff
@@ -22,6 +24,50 @@
 
 public abstract class Mapper implements ToXContentFragment, Iterable<Mapper> {
 
+    public static final NodeFeature STORE_SOURCE_MAPPER_PARAM = new NodeFeature("mapper.source.store_param");
+
+    static final String STORE_SOURCE_PARAM = "store_source";
+
+    // Only relevant for synthetic source mode.
+    public enum StoreSourceMode {
+        NONE("none"),      // No source recording
+        ARRAYS("arrays"),  // Store source for arrays of mapped fields
+        ALL("all");        // Store source for both singletons and arrays of mapped fields
+
+        StoreSourceMode(String name) {
+            this.name = name;
+        }
+
+        static StoreSourceMode fromString(String input) {
+            if (input.equals(NONE.name)) {
+                return NONE;
+            }
+            if (input.equals(ALL.name)) {
+                return ALL;
+            }
+            if (input.equals(ARRAYS.name)) {
+                return ARRAYS;
+            }
+            throw new IllegalArgumentException("Unknown " + STORE_SOURCE_PARAM + " value [" + input + "]");
+        }
+
+        @Override
+        public String toString() {
+            return name;
+        }
+
+        private final String name;
+    }
+
+    // Setting StoreSourceMode to ENABLED at the index level is equivalent to disabling synthetic source, which is not desired.
+    // Since the only valid option is to track array source by default, we use a boolean index setting for it.
+    public static final Setting<Boolean> STORE_ARRAY_SOURCE_SETTING = Setting.boolSetting(
```

Comment:
I am not really happy that the name will be different between index settings and mapper parameters (we have them named the same at least for `ignore_malformed` but i see why it is done. Can we maybe make `all` not valid for index setting but have the same name?

Reply:
Yeah that was my initial take here, too. An enum here is more flexible, we may want to have other values in the future.. 

@martijnvg wdyt?